plugin/session_view.py

```diff
@@ -119,6 +126,38 @@ def _apply_auto_complete_triggers(
         triggers.append(trigger)
         settings.set(self.AC_TRIGGERS_KEY, triggers)
 
+    def _do_code_lens_async(self) -> None:
+        if "codeLensProvider" in userprefs().disabled_capabilities:
+            return
+        if not self.has_capability_async("codeLensProvider"):
+            return
+        params = {"textDocument": text_document_identifier(self.view)}
+        self.session.send_request_async(Request("textDocument/codeLens", params, self.view), self._on_code_lens)
+
+    def _on_code_lens(self, response: Optional[List[CodeLens]]) -> None:
+        self.resolved_code_lenses = []
+        if not isinstance(response, list):
+            return
+        promises = []  # type: List[Promise[CodeLens]]
+        for code_lens in response:
+            if code_lens.get("command"):
+                self.resolved_code_lenses.append(code_lens)
+            else:
+                promises.append(self.session.send_request_task(Request("codeAction/resolve", code_lens, self.view)))
```

Comment:
`codeAction/resolve` -> `codeLens/resolve`
Hahahaha this is the reason why we got those `TypeError: 'Error' object is not subscriptable` :D

Reply:
After I changed that I can see the code actions in elm :)

![image](https://user-images.githubusercontent.com/22029477/99976872-432f6c80-2da4-11eb-8320-23972110aa15.png)

Reply:
I'm thinking we shouldn't use phantoms, but rather annotations for code lenses

- It would allow easier lazy resolvement of lenses (it can say [code lens available], and the user has to click on it, and only *then* do we do codeLense/resolve)
- Annotations don't make the text jump around unexpectedly

Reply:
this sounds like a bad UX for me because you would need to click to know what _is_ actually the code lens. So if your document as several code lenses you would spend time clicking around searching for the code lens you want.

Reply:
Okay. But the current code now tries to resolve *all* lenses when we get them. That means we could potentially make a language server try to do tens of find-reference queries, wasting CPU/time.

Ideally, these lenses are resolved once we see them in the viewport. But ST does not have a callback for when the viewport changed.

What we could do instead is to check where the selection is of the user and make a "tube" out of those selections and try to intersect those tubes with the regions of the unresolved lenses.

Reply:
> That means we could potentially make a language server try to do tens of find-reference queries, wasting CPU/time.

That's why Metals have settings to Opt-in and Optout of lenses on the fly.

Reply:
@ayoub-benali do you have some sort of tutorial for me to set up a Scala project with sbt? I am a total noob with this sbt build system.

Reply:
@rwols the easiest way to setup a project is to use the new command: https://www.scala-sbt.org/1.x/docs/sbt-new-and-Templates.html#Trying+new+command

After that you can edit or create scala files inside the scala directory.